### ISSUE TITLE
About reproducibility

### ISSUE THREAD

**Issue by helena-lena** [dwadden/multivers#9]
Hi, David:)
Thanks a lot for sharing your great work!

I am currently struggling on reproducing the results.
I tried finetuning on covidfact and scifact_20 dataset with one GPU(RTX 3090), but the results are different each time.

Looking at the metrics.csv file(in checkpoints folder) generated during finetuning, the label_loss, rationale_loss, and loss are recorded differently each time, so it seems that the randomness is not controlled during the training process.

When I looked at the code, I think there is no problem with the dataloader part because it is fixed. 
I tried to add the code (in anaconda3/envs/multivers/lib/python3.8/site-packages/pytorch_lightning/utilities/seed.py) for setting the seed as below, but it is still not reproducible.

![reproducibility](https://user-images.githubusercontent.com/72365222/223981885-86924434-8302-45aa-a58a-39962394dac9.png)

I wonder if there are any parts that need additional modification or if there are parts difficult to reproduce perfectly.
Thank you~~

**Comment by dwadden** [comment]
Thanks for raising this issue. I tried to control for randomness by calling [pl.seed_everything](https://github.com/dwadden/multivers/blob/main/multivers/train.py#L93) in the training script. Unfortunately, I think that GPU training is inherently [nondeterministic](https://pytorch.org/docs/stable/notes/randomness.html) for reasons that I don't understand.
Are you getting results within the same ballpark (within an F1 point or two) across runs, or are you getting wildly different results?

**Comment by helena-lena** [comment]
In the case of the covidfact dataset, I'm getting results within the same ballpark (within an F1 point or two), so that seems fine.
For the scifact_20 dataset, the results are quite different.
So I'm thinking of averaging multiple results.
Thanks for the quick response:)

**Comment by dwadden** [comment]
Hmm interesting, can you give me a sense of how different they are for scifact_20? I can try a few training runs and see if I get the same level of variation.

**Comment by helena-lena** [comment]
Yes:) For the scifact_20 dataset, the results are different as shown in the two figures below!

![scifact_1](https://user-images.githubusercontent.com/72365222/224584831-a53f882e-ceef-4a71-b007-2e8c77e83804.png)

![scifact_2](https://user-images.githubusercontent.com/72365222/224584837-2affeadc-66df-4845-b57a-9d41c414f563.png)

**Comment by dwadden** [comment]
Hmm so the results seem to vary by a few F1 points. Interesting. I'm traveling for the next couple weeks, but when I get back I'll try to take a look.

**Comment by dwadden** [comment]
Unfortunately I think I'm not going to have the bandwidth to run more experiments on this. I agree it's strange that there's so much variance between runs. If you have results on this, feel free to submit a Markdown document to the `doc` folder describing your findings and I can merge it in; that way we'll at least have this issue recorded.

**Comment by helena-lena** [comment]
Thank you for taking a look! Perhaps trying multiple experiments and averaging the results could be a potential solution.
